File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.py
class Solution(object):
    def is_possible(self, bloomDay, day, m , k):
        count =0;
        bouquets=0;
        for bloom in bloomDay:
            if bloom<=day:
               count+=1
               if count ==k:
                bouquets+=1 #one bouquet formed
                count=0
            else:
                count=0 #reset if a flower is not ready
        return bouquets>=m
    

    def minDays(self, bloomDay, m, k):
        """
        :type bloomDay: List[int]
        :type m: int
        :type k: int
        :rtype: int
        """
        if m*k>len(bloomDay):
            return -1 #not enough flowers
        low = min(bloomDay)
        high = max(bloomDay)
        answer =-1
        while low<=high:
            mid = (low+high)//2
            if self.is_possible(bloomDay, mid , m ,k):
                answer = mid #try to find smaller day
                high = mid -1 
            else:
                low = mid+1 #need more days 
        return answer
        

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    private static boolean isPossible(int[] bloomDay, int day,int m , int k){
        int count =0; // counts consecutive flowers that bloomed on or before 'day'
        int bouquets=0;
        for (int  bloom: bloomDay){
            if(bloom<=day){
                count++; //flower is ready
                if(count==k){
                    bouquets++; // form one bouquet
                    count = 0; // reset count for next bouquet
                }
            }else{
                    count = 0; // break in consecutive flowers
                }
            
        }
        return bouquets >= m; // check if required bouquets can be made

    }
    public int minDays(int[] bloomDay, int m, int k) {
        long required = (long)m*k; 
        if (required> bloomDay.length) return -1;//not enough flowers
        int minDay = Integer.MAX_VALUE;
        int maxDay = Integer.MIN_VALUE;

        // Find the minimum and maximum bloom day
        for (int bloom: bloomDay){
            minDay= Math.min(minDay, bloom);
            maxDay = Math.max(maxDay, bloom); 
        }
        // Binary search between minDay and maxDay
        int low= minDay, high = maxDay, result =-1;
        while(low<=high){
            int mid = (low +high)/2;
            if (isPossible(bloomDay, mid, m, k)) {
                result = mid; // possible to form bouquets, try earlier
                high = mid - 1;
            } else {
                low = mid + 1; // need more days
            }
        }
        return result;
    }
}